hail/src/main/scala/is/hail/io/fs/FS.scala

```diff
@@ -60,6 +73,114 @@ object BGZipCompressionCodec extends CompressionCodec {
   def makeOutputStream(os: OutputStream): OutputStream = new BGzipOutputStream(os)
 }
 
+object FSUtil {
+  def dropTrailingSlash(path: String): String = {
+    if (path.isEmpty)
+      return path
+
+    if (path.last != '/')
+      return path
+
+    var i = path.length - 1
+    while (i > 0 && path(i - 1) == '/')
+      i -= 1
+    path.substring(0, i)
+  }
+
+  def containsWildcard(path: String): Boolean = {
+    var i = 0
+    while (i < path.length) {
+      val c = path(i)
+      if (c == '\\') {
+        i += 1
+        if (i < path.length)
+          i += 1
+        else
+          return false
+      } else if (c == '*' || c == '{' || c == '?' || c == '[')
+        return true
+
+      i += 1
+    }
+
+    false
+  }
+}
+
+abstract class FSSeekableInputStream extends InputStream with Seekable {
+  protected[this] var closed: Boolean = false
+  protected[this] var pos: Long = 0
+  protected[this] var eof: Boolean = false
+
+  protected[this] val bb: ByteBuffer = ByteBuffer.allocate(64 * 1024)
+  bb.limit(0)
+
+  def fill(): Unit
+
+  override def read(): Int = {
+    if (eof)
+      return -1
+
+    if (bb.remaining() == 0) {
+      fill()
+      if (eof)
+        return -1
+    }
+
+    pos += 1
+    bb.get().toInt & 0xff
+  }
+
+  override def read(bytes: Array[Byte], off: Int, len: Int): Int = {
+    if (eof)
+      return -1
+
+    if (bb.remaining() == 0) {
+      fill()
+      if (eof)
+        return -1
+    }
+
+    val toTransfer = math.min(len, bb.remaining())
+    bb.get(bytes, off, toTransfer)
+    pos += toTransfer
+    toTransfer
```

Comment:
I believe this will result in a short read, and it's up to the callee to invoke `read` until `EOF`. Unclear to me which semantics we want but this is lifted from the current `GoogleStorageFS`.